# Riemann Hypothesis — Visualizations

This notebook contains all Riemann Hypothesis animations:

1. **Zeros on the Critical Line** — path of `ζ(½ + it)` in ℂ, with `|ζ|` vs `t`
2. **Critical Strip** — magnitude and phase heatmap of `ζ(s)` over `0 < Re(s) < 1`  
3. **Zeta Surface** — domain coloring of `ζ(s)` over the complex plane
4. **Winding Number** — argument principle: zeros enclosed = winding count
5. **Analytic Continuation** — Dirichlet series vs full `ζ(s)`

---
**The Riemann Hypothesis** states that all nontrivial zeros of `ζ(s)` lie on the **critical line** `Re(s) = ½`.

> *"The non-trivial zeros of ζ(s) have real part equal to 1/2."* — Bernhard Riemann, 1859

---

⚠️ **First run is slow** — zeta values are computed via mpmath and then cached. Subsequent runs are fast.

In [1]:
import matplotlib
matplotlib.rcParams['animation.embed_limit'] = 100  # MB

import mathplt.animations  # trigger auto-discovery
from mathplt.jupyter.widgets import AnimationWidget, list_animations
from mathplt.core.animator import AnimationConfig

list_animations()

Name,Description
complex,"Domain coloring of f(z) — hue=arg, brightness=|z|"
graph2d,"Animated 2D curve f(x, t) — equation input via EquationWidget"
graph3d,"Rotating 3D surface f(x, y)"
riemann.continuation,Analytic continuation of ζ(s) — Dirichlet series vs full zeta
riemann.critical_strip,Magnitude + phase heatmap of ζ(s) over the critical strip
riemann.winding,Argument principle — ζ(C) winding count equals zeros enclosed
riemann.zeros,Zeros of ζ(s) on the critical line — path in ℂ + |ζ| vs t
riemann.zeta_surface,Domain coloring of ζ(s) — full complex plane view


## 1. Zeros on the Critical Line

The path of `ζ(½ + it)` in the complex plane as `t` increases.
Each time the path **crosses the origin**, there is a zero of `ζ(s)`.
The Riemann Hypothesis says all nontrivial zeros satisfy `Re(s) = ½`.

In [ ]:
from mathplt.animations.riemann.zeros import RiemannZerosAnimator

config = AnimationConfig(
    fps=30,
    duration_seconds=20.0,
    figsize=(14, 6),
    dark_mode=True,
)
animator = RiemannZerosAnimator(
    config,
    t_min=0.5,
    t_max=50.0,
    n_zeros=15,
    resolution=2000,
)
AnimationWidget(animator).display()

Precomputing ζ(1/2 + it) values (will be cached after first run)...
  zeta_on_critical_line: 2000/2000...
Finding nontrivial zeros...
Found 15 zeros. First few t-values: ['14.135', '21.022', '25.011', '30.425', '32.935']


## 2. Critical Strip — Magnitude & Phase

Heatmap of `ζ(s)` over the critical strip `0 < Re(s) < 1`.
- **Left:** `log(1 + |ζ(s)|)` — dark spots along `Re(s) = ½` are zeros
- **Right:** `arg(ζ(s))` — phase structure shows how the function rotates

In [ ]:
from mathplt.animations.riemann.critical_strip import CriticalStripAnimator

config = AnimationConfig(fps=24, duration_seconds=15.0, figsize=(14, 7), dark_mode=True)
animator = CriticalStripAnimator(
    config,
    im_range=(0.5, 50.0),
    re_points=60,
    im_points=250,
)
AnimationWidget(animator).display()

## 3. Zeta Surface — Domain Coloring

Domain coloring of `ζ(s)` over the complex plane.
- White `+` markers: nontrivial zeros on the critical line
- Bright chaotic region at `s = 1`: the pole
- Structure to the left: trivial zeros at `s = -2, -4, -6, ...`

In [ ]:
from mathplt.animations.riemann.zeta_surface import ZetaSurfaceAnimator

config = AnimationConfig(fps=15, duration_seconds=8.0, figsize=(10, 10), dark_mode=True)
animator = ZetaSurfaceAnimator(
    config,
    re_range=(-1.5, 2.5),
    im_range=(0.5, 35.0),
    resolution=180,
    n_zeros_marked=10,
    animate_zoom=False,  # set True for slow zoom into critical strip
)
AnimationWidget(animator).display()

## 4. Winding Number — Argument Principle

The **argument principle** connects zeros to topology:

$$N_{\text{zeros inside C}} = \frac{1}{2\pi i} \oint_C \frac{\zeta'(s)}{\zeta(s)} ds = \text{winding number of } \zeta(C) \text{ around } 0$$

As the contour expands upward in Im(s), the image `ζ(C)` in the w-plane
winds around the origin once per enclosed zero.

In [ ]:
from mathplt.animations.riemann.winding_number import WindingNumberAnimator

config = AnimationConfig(fps=20, duration_seconds=20.0, figsize=(14, 7), dark_mode=True)
animator = WindingNumberAnimator(
    config,
    re_lo=0.1,
    re_hi=0.9,
    im_max=40.0,
    im_start=5.0,
    n_zeros=12,
)
AnimationWidget(animator).display()

## 5. Analytic Continuation

The Dirichlet series `Σ n^{-s}` only converges for `Re(s) > 1`.
The analytic continuation extends `ζ(s)` to the entire complex plane (except `s = 1`).

Watch how adding more terms fills in the domain coloring on the right half,
while the true `ζ(s)` (computed via mpmath) is always defined everywhere.

In [ ]:
from mathplt.animations.riemann.analytic_continuation import AnalyticContinuationAnimator

config = AnimationConfig(fps=15, duration_seconds=12.0, figsize=(14, 7), dark_mode=True)
animator = AnalyticContinuationAnimator(
    config,
    re_range=(-1.0, 3.0),
    im_range=(0.5, 25.0),
    resolution=120,
    max_terms=40,
)
AnimationWidget(animator).display()

---
## Background: The Riemann Zeta Function

**Definition (Dirichlet series, Re(s) > 1):**
$$\zeta(s) = \sum_{n=1}^{\infty} \frac{1}{n^s}$$

**Analytic continuation:** `ζ(s)` extends to all `s ∈ ℂ \ {1}` with a simple pole at `s = 1`.

**Functional equation:**
$$\zeta(s) = 2^s \pi^{s-1} \sin\left(\frac{\pi s}{2}\right) \Gamma(1-s) \zeta(1-s)$$

**Trivial zeros:** `s = -2, -4, -6, ...` (from the sine term)

**Nontrivial zeros:** All known zeros lie on `Re(s) = ½`. The Riemann Hypothesis says this is always true.

**First nontrivial zeros** (imaginary parts):
`14.135, 21.022, 25.011, 30.425, 32.935, 37.586, 40.919, 43.327, ...`